consensus/istanbul/backend/handler.go

```diff
@@ -286,3 +301,175 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
 		}
 	}
 }
+
+// Handshake allows the initiating peer to identify itself as a validator
+func (sb *Backend) Handshake(peer consensus.Peer) (bool, error) {
+	// Only written to if there was a non-nil error when sending or receiving
+	errCh := make(chan error, 1)
+	isValidatorCh := make(chan bool, 1)
+
+	sendHandshake := func() {
+		var validatorProofMsg *istanbul.Message
+		var err error
+		containsProof := false
+		if peer.PurposeIsSet(p2p.ValidatorPurpose) {
+			// validatorProofMsg may be nil
+			validatorProofMsg, err = sb.retrieveVersionedEnodeMsg()
+			if err != nil {
+				errCh <- err
+				return
+			}
+		}
+		// Even if we decide not to identify ourselves,
+		// send an empty message to complete the handshake
+		if validatorProofMsg == nil {
+			validatorProofMsg = &istanbul.Message{}
+		} else {
+			containsProof = true
+		}
+		msgBytes, err := validatorProofMsg.Payload()
+		if err != nil {
+			errCh <- err
+			return
+		}
+		err = peer.Send(istanbulValidatorProofMsg, msgBytes)
+		if err != nil {
+			errCh <- err
+		}
+		isValidatorCh <- containsProof
+	}
+	readHandshake := func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		isValidatorCh <- isValidator
+	}
+
+	// Only the initating peer sends the message
+	if peer.Inbound() {
+		go readHandshake()
+	} else {
+		go sendHandshake()
+	}
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	for {
+		select {
+		case err := <-errCh:
+			return false, err
+		case <-timeout.C:
+			return false, p2p.DiscReadTimeout
+		case isValidator := <-isValidatorCh:
+			return isValidator, nil
+		}
+	}
+}
+
+// shouldSendValidatorProof determines if this node should send a
+// validator proof revealing its address to a peer
+func (sb *Backend) shouldSendValidatorProof(peer consensus.Peer) (bool, error) {
+	var validatorAddress common.Address
+	if sb.config.Proxy {
+		validatorAddress = sb.config.ProxiedValidatorAddress
+	} else {
+		validatorAddress = sb.Address()
+	}
+	// Check to see if this node is a validator
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return false, err
+	}
+	if !regAndActiveVals[validatorAddress] {
+		return false, nil
+	}
+	// If the peer is not a known validator, we do not want to give up extra
+	// information about ourselves
+	if _, err := sb.valEnodeTable.GetAddressFromNodeID(peer.Node().ID()); err != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	logger := sb.logger.New("func", "readValidatorProofMessage")
+	peerMsg, err := peer.ReadMsg()
```

Comment:
Based on how low level `ReadMsg` is, we should verify the message code and size before decoding, much as the `peer.readStatus` method does. https://github.com/celo-org/celo-blockchain/blob/b7a7de550837eeb8ac1b30aa0f9baf4721c7d3a4/eth/peer.go#L454-L468